docs/architecture/adr-010.md

```diff
@@ -0,0 +1,138 @@
+
+## Changelog
+* 2023-10-02: first draft
+
+## Context
+
+Cronos is an EVM-compatible built on Cosmos SDK. The "ethermint" module allows a full compatibility and interoperability with Ethereum. However, it comes with the sacrifice of cosmos native features or cosmos interchain features (IBC)
+as end-users are not able to access them through an ethereum transaction format. The solution is to provide precompile, allowing users to access to cosmos features through smart contracts. 
+A precompiled behave like a smart contract, it is designed by an address and can be invoked by other smart contracts. The difference is that its execution is not inside a smart contract.
+App-chain specific logics can be embedded to the pre-compiled so that it could be used to expose cosmos native features to the EVM layer. 
+
+
+## Decision
+
+### Fork go-ethereum for stateful precompiled support
+
+One of the major difficulties in supporting pre-compiled is the lack of flexibility in geth for defining custom precompiled and its non-ability to modify the state.
+We made the choice of forking geth and implement additional interfaces and logics so that app-chain can define its own precompiled.
+
+The PR [#7](https://github.com/evmos/go-ethereum/pull/7) implements the ability to set custom precompiled to the EVM.
+while this PR [#10](https://github.com/evmos/go-ethereum/pull/10) allows the precompiled to modify the state (for stateful precompiled)
+
+In the future, we are planning to maintain the following fork permanently
+`github.com/evmos/go-ethereum` 
+
+### Implement pre-compiled for cosmos native module
+
+We provide pre-compile for cosmos native module such as bank or staking.
+
+The input and output of methods are mapped with those defined cosmos native transactions
+
+```solidity
+
+interface IBankModule {
+    function mint(address,uint256) external payable returns (bool);
+    function balanceOf(address,address) external view returns (uint256);
+    function burn(address,uint256) external payable returns (bool);
+    function transfer(address,address,uint256) external payable returns (bool);
+}
+
+
+
+interface IStakingModule {
+    # to be completed
+}
+```
+
+To call a precompiled, create an instance of the module interface with the precompiled address. 
+The methods are ABI encoded, and can be called directly as calling the function through the interface:
+
+
+```solidity
+
+address constant bankContract = 0x0000000000000000000000000000000000000064;
+IBankModule bank = IBankModule(bankContract);
+bank.transfer(msg.sender, recipient, amount);
+ 
+```
+
+
+### Implement pre-compiled for cosmos interchain communication
+
+#### Relayer
+One of the benefits of supporting precompiled for IBC is that it would be possible to track the change of state triggered by the ibc transaction with ethereum receipts.
+
+For that reason we are exposing all relayer related actions to the evm layer. An action is defined by a 4 bytes prefix. 
+
+```go
+
+const (
```

Comment:
@mmsqe Not sure why we are using a prefix based encoding rather than abi encoding like for all the other precompiled?